### PR TITLE
Charm uses charmed-mongodb snap

### PR DESCRIPTION
**Pull request by MiaAltieri**
## What is this PR?
This PR is a replacement of the old apt install code with the snap code. This involved quite a few changes:
1. how and where files were written
2. updates of how `mongosh` was used (we now use `charmed-mongodb.mongo`)
3. updates to docs 
4. how mongod was started/stopped
5. removal of un-used code

## Tests
Integration + Unit tests updated 

## Follow up PR(s)
1. we will remove the `percona-backup-mongodb` charm and verify that the charm can support backups with the "fat"-snap installation of `pbm`
2. Change release to `22.04` and verify functionality 
3. Auto update the release of the snap and copy over data from the snap_data folder when a new release is made

### CODE REVIEW

File: lib/charms/mongodb/v0/helpers.py [hunk 1]
```diff
@@ -64,7 +66,7 @@ def get_create_user_cmd(config: MongoDBConfiguration, mongo_path="mongo") -> Lis
     ]
 
 
-def get_mongod_cmd(config: MongoDBConfiguration) -> str:
+def get_mongod_cmd(config: MongoDBConfiguration, mongo_files_dir="/etc/mongodb") -> str:
```
Comment (jardon) [inline lib/charms/mongodb/v0/helpers.py:69]: Is this path correct?  Shouldn't this be under common?

File: documentation/tutorial.md [hunk 2]
```diff
@@ -774,7 +774,7 @@ After `ssh`ing into `mongodb/0`, we are now in the unit that is hosting Charmed
 
 As before, we will connect to MongoDB via the saved MongoDB URI. Connect using the saved URI and the following TLS options:
 ```shell
-mongosh mongodb://$DB_USERNAME:$DB_PASSWORD@$HOST_IP/$DB_NAME?replicaSet=$REPL_SET_NAME --tls --tlsCAFile /etc/mongodb/external-ca.crt --tlsCertificateKeyFile /etc/mongodb/external-cert.pem
+charmed-mongodb.mongo mongodb://$DB_USERNAME:$DB_PASSWORD@$HOST_IP/$DB_NAME?replicaSet=$REPL_SET_NAME --tls --tlsCAFile /etc/mongodb/external-ca.crt --tlsCertificateKeyFile /etc/mongodb/external-cert.pem
```
Comment (jardon) [inline documentation/tutorial.md:777]: Similar to the other comment, should these paths be updated?  The snap shouldnt be able to access that directory.

**Review by jardon — COMMENTED** [review]
Have a couple questions for ya, but otherwise looks great!

File: documentation/tutorial.md (hunk 2 above)
Comment (MiaAltieri) [reply documentation/tutorial.md:777]: yes! great eyes thank you Jarred :)

File: lib/charms/mongodb/v0/helpers.py (hunk 1 above)
Comment (MiaAltieri) [reply lib/charms/mongodb/v0/helpers.py:69]: that is the default directory, it defaults to `etc/mongodb` so that the k8s charm won't require any changes when the lib gets updated

**Review by jardon — APPROVED** [review]

File: src/charm.py [hunk 3]
```diff
@@ -74,7 +60,10 @@
 
 # We expect the MongoDB container to use the default ports
 MONGODB_PORT = 27017
-SNAP_PACKAGES = [("percona-backup-mongodb", "edge")]
+SNAP_PACKAGES = [
+    ("percona-backup-mongodb", "edge"),
+    ("charmed-mongodb", "5.0/edge/pass-args"),
```
Comment (MiaAltieri) [inline src/charm.py:65]: change this before merging

File: src/charm.py [hunk 4]
```diff
@@ -298,20 +287,20 @@ def _on_mongodb_relation_handler(self, event: ops.charm.RelationEvent) -> None:
...
-        """
+        """Handle the install event (fired on startup)."""
         self.unit.status = MaintenanceStatus("installing MongoDB")
         try:
-            self._add_repository(REPO_URL, GPG_URL, REPO_ENTRY)
-            self._install_apt_packages(["mongodb-org"])
             self._install_snap_packages(packages=SNAP_PACKAGES)
-        except (apt.InvalidSourceError, ValueError, apt.GPGKeyError, URLError, snap.SnapError):
+        except snap.SnapError:
             self.unit.status = BlockedStatus("couldn't install MongoDB")
```
Comment (MiaAltieri) [inline src/charm.py:295]: return on failure

File: metadata.yaml [hunk 5]
```diff
@@ -20,7 +20,8 @@ provides:
 storage:
   mongodb:
     type: filesystem
-    location: /data/db
+    # mongod is strictly confined in the charmed-mongodb snap
+    location: /var/snap/charmed-mongodb/common/db
```
Comment (delgod) [inline metadata.yaml:24]: Location in snap should match the spec - https://docs.google.com/document/d/1MyT4Wt_-vgPwhwEjC-oeHRdLYrZWaQzFGTwclT7hD1U/edit

File: src/machine_helpers.py [hunk 6]
```diff
@@ -170,12 +113,16 @@ def generate_service_args(auth: bool, machine_ip: str, config: MongoDBConfigurat
...
     Path(parent_dir).mkdir(parents=True, exist_ok=True)
+    file_name = f"{parent_dir}/{file_name}"
     with open(file_name, "w") as write_file:
         write_file.write(file_contents)
 
-    os.chmod(file_name, 0o400)
+    if "keyFile" in file_name:
+        os.chmod(file_name, 0o400)
+    else:
+        os.chmod(file_name, 0o440)
```
Comment (delgod) [inline src/machine_helpers.py:119]: I think there some mistake
```suggestion
    if os.path.exists(file_name):
        os.chmod(file_name, 0o400)
```

File: src/machine_helpers.py [hunk 7]
```diff
@@ -18,102 +18,44 @@
...
+    with open(ENV_VAR_PATH, "w") as service_file:
+        service_file.writelines(env_vars)
 
     # mongod requires permissions to /data/db
     mongodb_user = pwd.getpwnam(MONGO_USER)
     os.chown(MONGO_DATA_DIR, mongodb_user.pw_uid, mongodb_user.pw_gid)
 
-    # changes to service files are only applied after reloading
+    # changes to service files are only applied after reloading, this is needed even for snaps
     systemd.daemon_reload()
```
Comment (delgod) [inline src/machine_helpers.py:52]: Daemon reload is not needed if changes are in the env file.

**Review by delgod — CHANGES_REQUESTED** [review]

File: src/charm.py [hunk 8]
```diff
@@ -298,20 +287,21 @@ def _on_mongodb_relation_handler(self, event: ops.charm.RelationEvent) -> None:
...
+            return
 
         # if a new unit is joining a cluster with a legacy relation it should start without auth
         auth = not self.client_relations._get_users_from_relations(None, rel="obsolete")
+
+        # clear the default config file - user provided config files will be added in the config
+        # hook
+        check_output("truncate -s 0 /var/snap/charmed-mongodb/common/mongod.conf", shell=True)
```
Comment (delgod) [inline src/charm.py:303]: Why not do this with python, why shell needed?

File: src/charm.py [hunk 9]
```diff
@@ -371,15 +364,20 @@ def _on_update_status(self, event):
...
-                # ensure that the correct port is open for the service
-                self._open_port_tcp(self._port)
-                self.unit.status = WaitingStatus("Waiting for MongoDB to start")
-                event.defer()
-                return
+        try:
+            with MongoDBConnection(self.mongodb_config, "localhost", direct=True) as direct_mongo:
+                if not direct_mongo.is_ready:
+                    logger.debug("mongodb service is not ready yet, restarting.")
+                    self.restart_mongod_service()
```
Comment (delgod) [inline src/charm.py:371]: I don't think it is needed anymore if the snap unit has `restart-condition=always` and `restart-delay=1m`

File: metadata.yaml (hunk 5 above)
Comment (MiaAltieri) [reply metadata.yaml:24]: Also I think the config file should change as well

File: src/machine_helpers.py (hunk 6 above)
Comment (MiaAltieri) [reply src/machine_helpers.py:115]: resolved per our conversation in mattermost. 

TL;DR is that it is necessary that the KeyFile has `400` for mongod to start, but for files like the TLS files it is necessary to have `440` so the user can verify TLS ops

File: src/charm.py [hunk 10]
```diff
@@ -74,7 +65,10 @@
 
 # We expect the MongoDB container to use the default ports
 MONGODB_PORT = 27017
-SNAP_PACKAGES = [("percona-backup-mongodb", "edge")]
+SNAP_PACKAGES = [
+    ("percona-backup-mongodb", "edge"),
+    ("charmed-mongodb", "5/edge/update-file-paths"),
```
Comment (MiaAltieri) [inline src/charm.py:70]: update this to 5/edge before pushing

File: src/charm.py [hunk 11]
```diff
@@ -298,20 +292,22 @@ def _on_mongodb_relation_handler(self, event: ops.charm.RelationEvent) -> None:
...
+            return
 
         # if a new unit is joining a cluster with a legacy relation it should start without auth
         auth = not self.client_relations._get_users_from_relations(None, rel="obsolete")
+
+        # clear the default config file - user provided config files will be added in the config
+        # changed hook
+        with open(MONGOD_CONF_FILE_PATH, "r+") as file:
+            file.truncate(0)
+
```
Comment (delgod) [inline src/charm.py:310]:
```suggestion
        try:
            with open(MONGOD_CONF_FILE_PATH, 'r+') as f:
                f.truncate(0)
         except IOError:
             self.unit.status = BlockedStatus("couldn't install MongoDB")
             return
```

File: src/machine_helpers.py [hunk 12]
```diff
@@ -14,104 +14,45 @@
...
-        return
-
-    logger.debug("starting mongod.service")
-    try:
-        systemd.service_start("mongod.service")
-    except systemd.SystemdError as e:
-        logger.error("failed to enable mongod.service, error: %s", str(e))
-        raise
+MONGOD_DATA_DIR = f"{MONGODB_COMMON_DIR}/var/lib/mongod"
+MONGODB_DB_DIR = f"{MONGOD_DATA_DIR}/db"
```
Comment (delgod) [inline src/machine_helpers.py:32]:
```suggestion
MONGODB_DB_DIR = f"{MONGOD_DATA_DIR}"
```

File: metadata.yaml [hunk 13]
```diff
@@ -20,7 +20,8 @@ provides:
 storage:
   mongodb:
     type: filesystem
-    location: /data/db
+    # mongod is strictly confined in the charmed-mongodb snap
+    location: /var/snap/charmed-mongodb/common/var/lib/mongod/db
```
Comment (delgod) [inline metadata.yaml:24]:
```suggestion
    location: /var/snap/charmed-mongodb/common/var/lib/mongod
```

File: src/machine_helpers.py (hunk 6 above)
Comment (delgod) [reply src/machine_helpers.py:122]: Can we add a code comment here? to explain the because of MongoDB limitation it is needed 400 rights for keyfile and we need 440 rights on tls certs to be able to connect via MongoDB shell

**Review by delgod — CHANGES_REQUESTED** [review]

File: src/machine_helpers.py (hunk 12 above)
Comment (jardon) [reply src/machine_helpers.py:32]: We don't want the db directory to be separate? If that's the case, I can remove that entire constant and just use `MONGOD_DATA_DIR`

File: metadata.yaml [hunk 14]
```diff
@@ -20,7 +20,8 @@ provides:
 storage:
   mongodb:
     type: filesystem
-    location: /data/db
+    # mongod is strictly confined in the charmed-mongodb snap
+    location: /var/snap/charmed-mongodb/common/var/lib/mongod
```
Comment (jardon) [inline metadata.yaml:24]: Would it make more sense for this to be `/var/snap/charmed-mongodb/common/var/lib/mongodb` instead to mirror the deb package?  It would essentially be `$SNAP_COMMON/var/lib/mongodb` where the deb is `/var/lib/mongodb`

**Review by jardon — CHANGES_REQUESTED** [review]

File: metadata.yaml (hunk 14 above)
Comment (delgod) [reply metadata.yaml:24]: lets use `/var/lib/mongodb`, but for data dir only.
config dir should remain the same

**Review by delgod — APPROVED** [review]

File: lib/charms/mongodb/v0/helpers.py [hunk 15]
```diff
@@ -23,18 +23,20 @@
```
Comment (delgod) [inline lib/charms/mongodb/v0/helpers.py:23]: Can you please increase LIBPATCH?

File: lib/charms/mongodb/v0/helpers.py [hunk 16]
```diff
@@ -64,7 +66,7 @@ def get_create_user_cmd(config: MongoDBConfiguration, mongo_path="mongo") -> Lis
     ]
 
 
-def get_mongod_cmd(config: MongoDBConfiguration) -> str:
+def get_mongod_cmd(config: MongoDBConfiguration, mongo_files_dir="/etc/mongodb") -> str:
     """Construct the MongoDB startup command line.
 
     Returns:
```
Comment (delgod) [inline lib/charms/mongodb/v0/helpers.py:72]: can you please add `--dbpath` arg with path to DATA_DIR? I want to have it passed via arg

**Review by taurus-forever — APPROVED** [review]
Nice and well done!

File: src/charm.py [hunk 17]
```diff
@@ -74,7 +65,10 @@
 
 # We expect the MongoDB container to use the default ports
 MONGODB_PORT = 27017
-SNAP_PACKAGES = [("percona-backup-mongodb", "edge")]
+SNAP_PACKAGES = [
+    ("percona-backup-mongodb", "edge"),
```
[the remark this answers is not in the document]
Comment (MiaAltieri) [reply src/charm.py:69]: Hi yes! Please see the PR description
> Follow up PR(s)
we will remove the percona-backup-mongodb charm and verify that the charm can support backups with the "fat"-snap installation of pbm

**Review by delgod — APPROVED** [review]
